worlds/civ_6/Regions.py

```diff
@@ -0,0 +1,175 @@
+        era, True, item_table)
+    if not state.has_all([item.name for item in all_previous_items_no_progressives], player):
+        return False
+
+    # Verify we have the correct amount of progressive items
+    all_previous_items = get_prereqs_for_era(
+        era, False, item_table)
+    required_counts: Dict[str, int] = {}
+
+    for key, value in progressive_districts.items():
+        required_counts[key] = 0
+        for item in all_previous_items:
+            if item.civ_name in value:
+                required_counts[key] += 1
+
+    return state.has_all_counts({format_item_name(key): value for key, value in required_counts.items()}, player)
+
+
+def has_required_progressive_eras(state: CollectionState, era: EraType, player: int) -> bool:
+    """Checks, for the given era, how many are required to proceed to the next era. Ancient = 0, Classical = 1, etc."""
+    if era == EraType.ERA_FUTURE or era == EraType.ERA_INFORMATION:
+        return True
+
+    eras = [e.value for e in EraType]
+    era_index = eras.index(era.value)
+    return state.has(format_item_name("PROGRESSIVE_ERA"), player, era_index + 1)
+
+
+def has_required_items(state: CollectionState, era: EraType, world: 'CivVIWorld') -> bool:
+    player = world.player
+    has_progressive_districts = world.options.progression_style != "none"
+    has_progressive_eras = world.options.progression_style == "eras_and_districts"
+
+    if has_progressive_districts:
+        required_items = has_required_progressive_districts(state, era, player)
+    else:
+        era_required_items = [get_item_by_civ_name(item, world.item_table).name for item in get_era_required_items_data()[era.value]]
+        required_items = state.has_all(era_required_items, player)
+
+    return required_items and (not has_progressive_eras or has_required_progressive_eras(state, era, player))
+
+
+def create_regions(world: 'CivVIWorld'):
+    menu = Region("Menu", world.player, world.multiworld)
+    world.multiworld.regions.append(menu)
+
+    has_progressive_eras = world.options.progression_style == "eras_and_districts"
+    has_goody_huts = world.options.shuffle_goody_hut_rewards
+    has_boosts = world.options.boostsanity
+
+    regions: List[Region] = []
+    for era in EraType:
+        era_region = Region(era.value, world.player, world.multiworld)
+        era_locations: Dict[str, Optional[int]] = {location.name: location.code for _key,
+                                                   location in world.location_by_era[era.value].items()}
+
+        if not has_progressive_eras:
+            era_locations = {key: value for key, value in era_locations.items() if key.split("_")[0] != "ERA"}
+        if not has_goody_huts:
+            era_locations = {key: value for key, value in era_locations.items() if key.split("_")[0] != "GOODY"}
+        if not has_boosts:
+            era_locations = {key: value for key, value in era_locations.items() if key.split("_")[0] != "BOOST"}

```

Comment:
Are these intended to be mutually exclusive like this?
It looks like if you have goody huts and boosts off, when it gets to the `if not has_boosts` then it'll add back the goody hut locations?

Reply:
Ohhh, okay yeah that makes way more sense then

Reply:
Good eye, I think I'm actually just going to rework this a little bit. It is hard to read and unnecessarily confusing 👍

Reply:
I ended up making this a lot easier to read and so it only had one loop per era rather than 4
```
    for era in EraType:
        era_region = Region(era.value, world.player, world.multiworld)
        era_locations: Dict[str, Optional[int]] = {}

        for key, location in world.location_by_era[era.value].items():
            category = key.split("_")[0]
            if optional_location_inclusions.get(category, True):
                era_locations[location.name] = location.code

        era_region.add_locations(era_locations, CivVILocation)

        regions.append(era_region)
        world.multiworld.regions.append(era_region)
  ```